### В первом мини-проекте вам предстоит проанализировать данные о бронировании отелей. 


 

-  Задачи
1. Импортируйте библиотеку pandas как pd. Загрузите датасет bookings.csv с разделителем ;. Проверьте размер таблицы, типы          переменных, а затем выведите первые 7 строк, чтобы посмотреть на данные. 

2. Приведите названия колонок к нижнему регистру и замените пробелы на знак нижнего подчеркивания.

3. Пользователи из каких стран совершили наибольшее число успешных бронирований? Укажите топ-5.

4. На сколько ночей в среднем бронируют отели разных типов?

5. Иногда тип номера, полученного клиентом (assigned_room_type), отличается от изначально забронированного (reserved_room_type).    Такое может произойти, например, по причине овербукинга. Сколько подобных наблюдений встретилось в датасете?

6. Проанализируйте даты запланированного прибытия. – На какой месяц чаще всего успешно оформляли бронь в 2016? Изменился ли        самый популярный месяц в 2017?– Сгруппируйте данные по годам и проверьте, на какой месяц бронирования отеля типа City Hotel      отменялись чаще всего в каждый из периодов

7. Посмотрите на числовые характеристики трёх переменных: adults, children и babies. Какая из них имеет наибольшее среднее          значение?

8. Создайте колонку total_kids, объединив children и babies. Для отелей какого типа среднее значение переменной оказалось          наибольшим?

9. Создайте переменную has_kids, которая принимает значение True, если клиент при бронировании указал хотя бы одного ребенка        (total_kids), в противном случае – False. Посчитайте отношение количества ушедших пользователей к общему количеству клиентов,    выраженное в процентах (churn rate). Укажите, среди какой группы показатель выше.

## Описание данных
   
   Имеются следующие переменные:

- Hotel – тип отеля (City Hotel или Resort Hotel)  
- Is canceled – бронирование было отменено (1) или нет (0); не отмененное считается успешным
- Lead time – количество дней, прошедших между датой бронирования и датой прибытия  
- Arrival full date – полная дата прибытия
- Arrival date year – год прибытия  
- Arrival date month – месяц прибытия  
- Arrival date week number – номер недели прибытия
- Arrival date day of month – день прибытия
- Stays in weekend nights – количество выходных (суббота или воскресенье), которые гость забронировал для проживания в отеле
- Stays in week nights – количество дней (с понедельника по пятницу), которые гость забронировал для проживания в отеле
- Stays total nights – общее число забронированных ночей (сумма двух предыдущих колонок)
- Adults – число взрослых
- Children – число детей
- Babies – число младенцев 
- Meal – выбранный тип питания
- Country – страна происхождения клиента
- Reserved room type – тип зарезервированного номера
- Assigned room type – тип полученного номера (может отличаться от забронированного)
- Customer type – тип бронирования
- Reservation status – значение последнего статуса брони: Canceled - было отменено клиентом; Check-Out - клиент            зарегистрировался, но уже покинул отель; No-Show - клиент не зарегистрировался и сообщил администрации отеля причину
- Reservation status date – дата обновления статуса

____

## Задание 1
Импортируйте библиотеку pandas как pd. Загрузите датасет bookings.csv с разделителем ;. Проверьте размер таблицы, типы   переменных, а затем выведите первые 7 строк, чтобы посмотреть на данные.

In [24]:
import pandas as pd 
import numpy as np

In [3]:
bookings = pd.read_csv('C://Users/рома/Downloads/2_bookings.csv', sep = ';')
bookings.head(7)

Hotel  Is Canceled  Lead Time arrival full date  Arrival Date Year  \
0  Resort Hotel            0        342        2015-07-01               2015   
1  Resort Hotel            0        737        2015-07-01               2015   
2  Resort Hotel            0          7        2015-07-01               2015   
3  Resort Hotel            0         13        2015-07-01               2015   
4  Resort Hotel            0         14        2015-07-01               2015   
5  Resort Hotel            0         14        2015-07-01               2015   
6  Resort Hotel            0          0        2015-07-01               2015   

  Arrival Date Month  Arrival Date Week Number  Arrival Date Day of Month  \
0               July                        27                          1   
1               July                        27                          1   
2               July                        27                          1   
3               July                        27                          1   
4               July                        27                          1   
5               July                        27                          1   
6               July                        27                          1   

   Stays in Weekend nights  Stays in week nights  ...  Adults  Children  \
0                        0                     0  ...       2       0.0   
1                        0                     0  ...       2       0.0   
2                        0                     1  ...       1       0.0   
3                        0                     1  ...       1       0.0   
4                        0                     2  ...       2       0.0   
5                        0                     2  ...       2       0.0   
6                        0                     2  ...       2       0.0   

   Babies  Meal Country Reserved Room Type Assigned room type customer type  \
0       0    BB     PRT                  C                  C     Transient   
1       0    BB     PRT                  C                  C     Transient   
2       0    BB     GBR                  A                  C     Transient   
3       0    BB     GBR                  A                  A     Transient   
4       0    BB     GBR                  A                  A     Transient   
5       0    BB     GBR                  A                  A     Transient   
6       0    BB     PRT                  C                  C     Transient   

  Reservation Status Reservation status_date  
0          Check-Out              2015-07-01  
1          Check-Out              2015-07-01  
2          Check-Out              2015-07-02  
3          Check-Out              2015-07-02  
4          Check-Out              2015-07-03  
5          Check-Out              2015-07-03  
6          Check-Out              2015-07-03  

[7 rows x 21 columns]

In [4]:
bookings.shape

(119390, 21)

In [5]:
bookings.dtypes

Hotel                         object
Is Canceled                    int64
Lead Time                      int64
arrival full date             object
Arrival Date Year              int64
Arrival Date Month            object
Arrival Date Week Number       int64
Arrival Date Day of Month      int64
Stays in Weekend nights        int64
Stays in week nights           int64
stays total nights             int64
Adults                         int64
Children                     float64
Babies                         int64
Meal                          object
Country                       object
Reserved Room Type            object
Assigned room type            object
customer type                 object
Reservation Status            object
Reservation status_date       object
dtype: object

____

## Задание 2
Приведите названия колонок к нижнему регистру и замените пробелы на знак нижнего подчеркивания.

In [6]:
bookings.columns

Index(['Hotel', 'Is Canceled', 'Lead Time', 'arrival full date',
       'Arrival Date Year', 'Arrival Date Month', 'Arrival Date Week Number',
       'Arrival Date Day of Month', 'Stays in Weekend nights',
       'Stays in week nights', 'stays total nights', 'Adults', 'Children',
       'Babies', 'Meal', 'Country', 'Reserved Room Type', 'Assigned room type',
       'customer type', 'Reservation Status', 'Reservation status_date'],
      dtype='object')

In [10]:
# создадим небольшую фуекцию которая принимает на вход один аргумент 
# и возвращает значение этого аргумента в нижнем регистре и без пробела 
def replace_space(name):
    new_name = name.replace(' ', '_').lower()
    return new_name

In [11]:
# переименовываем наши колонки с помощью метода rename и нашей функции
bookings = bookings.rename(columns = replace_space)
bookings.head()

hotel  is_canceled  lead_time arrival_full_date  arrival_date_year  \
0  Resort Hotel            0        342        2015-07-01               2015   
1  Resort Hotel            0        737        2015-07-01               2015   
2  Resort Hotel            0          7        2015-07-01               2015   
3  Resort Hotel            0         13        2015-07-01               2015   
4  Resort Hotel            0         14        2015-07-01               2015   

  arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  \
0               July                        27                          1   
1               July                        27                          1   
2               July                        27                          1   
3               July                        27                          1   
4               July                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  ...  adults  children  \
0                        0                     0  ...       2       0.0   
1                        0                     0  ...       2       0.0   
2                        0                     1  ...       1       0.0   
3                        0                     1  ...       1       0.0   
4                        0                     2  ...       2       0.0   

   babies  meal country reserved_room_type assigned_room_type customer_type  \
0       0    BB     PRT                  C                  C     Transient   
1       0    BB     PRT                  C                  C     Transient   
2       0    BB     GBR                  A                  C     Transient   
3       0    BB     GBR                  A                  A     Transient   
4       0    BB     GBR                  A                  A     Transient   

  reservation_status reservation_status_date  
0          Check-Out              2015-07-01  
1          Check-Out              2015-07-01  
2          Check-Out              2015-07-02  
3          Check-Out              2015-07-02  
4          Check-Out              2015-07-03  

[5 rows x 21 columns]

____________

## Задание 3
Пользователи из каких стран совершили наибольшее число успешных бронирований? Укажите топ-5.

In [12]:
bookings[bookings['is_canceled']==0]['country'].value_counts().head(5)

PRT    21071
GBR     9676
FRA     8481
ESP     6391
DEU     6069
Name: country, dtype: int64

___

## Задание 4
На сколько ночей в среднем бронируют отели разных типов?

In [13]:
round(bookings.groupby('hotel',as_index=False)\
              .agg({'stays_total_nights':'mean'}),2)

hotel  stays_total_nights
0    City Hotel                2.98
1  Resort Hotel                4.32

____

## Задание 5

Иногда тип номера, полученного клиентом (assigned_room_type), отличается от изначально забронированного (reserved_room_type). Такое может произойти, например, по причине овербукинга. Сколько подобных наблюдений встретилось в датасете?

In [14]:
bookings.query(" assigned_room_type != reserved_room_type ").shape

(14917, 21)

В 14917 случаях тип номера отличался от изначально забронированного 

___

## Задание 6
Проанализируйте даты запланированного прибытия. – На какой месяц чаще всего успешно оформляли бронь в 2016? Изменился ли самый популярный месяц в 2017?
Сгруппируйте данные по годам и проверьте, на какой месяц бронирования отеля типа City Hotel отменялись чаще всего в каждый из периодов

In [19]:
bookings.groupby('arrival_date_year')\
        .arrival_date_month\
        .agg(pd.Series.mode)

arrival_date_year
2015    September
2016      October
2017          May
Name: arrival_date_month, dtype: object

c помощью группировки и моды выяснили что самый популярный месяц в 2016 - октябрь а в 2017 - май.

In [20]:
bookings.query("hotel == 'City Hotel'")\
        .groupby(['arrival_date_year','arrival_date_month'])\
        .is_canceled.sum()\
        .groupby('arrival_date_year')\
        .nlargest(1)

arrival_date_year  arrival_date_year  arrival_date_month
2015               2015               September             1543
2016               2016               October               1947
2017               2017               May                   2217
Name: is_canceled, dtype: int64

Чаще всего отменялись бронирования в сентябре в 2015, в октябре в 2016 и в мае в 2017

___

## Задание 7
Посмотрите на числовые характеристики трёх переменных: adults, children и babies. Какая из них имеет наибольшее среднее значение?

In [21]:
bookings[['adults','children','babies']].describe()

adults       children         babies
count  119390.000000  119386.000000  119390.000000
mean        1.856403       0.103890       0.007949
std         0.579261       0.398561       0.097436
min         0.000000       0.000000       0.000000
25%         2.000000       0.000000       0.000000
50%         2.000000       0.000000       0.000000
75%         2.000000       0.000000       0.000000
max        55.000000      10.000000      10.000000

наибольшее среднее значение имеет колонка adults

## Задание 8
Создайте колонку total_kids, объединив children и babies. Для отелей какого типа среднее значение переменной оказалось наибольшим?

In [22]:
bookings['total_kids'] = bookings['children'] + bookings['babies']

In [23]:
bookings.groupby('hotel')['total_kids'].mean()

hotel
City Hotel      0.096311
Resort Hotel    0.142586
Name: total_kids, dtype: float64

для отелей типа resort среднее оказалось больше

## Задание 9
Создайте переменную has_kids, которая принимает значение True, если клиент при бронировании указал хотя бы одного ребенка (total_kids), в противном случае – False. Посчитайте отношение количества ушедших пользователей к общему количеству клиентов, выраженное в процентах (churn rate). Укажите, среди какой группы показатель выше.

In [29]:
bookings['has_kids'] = bookings['total_kids'] >= 1 
bookings['has_kids'].value_counts()

False    110058
True       9332
Name: has_kids, dtype: int64

In [30]:
bookings[['has_kids','is_canceled']].value_counts()


has_kids  is_canceled
False     0              69093
          1              40965
True      0               6073
          1               3259
dtype: int64

In [41]:
rates = bookings.groupby('has_kids')['is_canceled'].value_counts(normalize=True)
round(rates*100,2)

has_kids  is_canceled
False     0              62.78
          1              37.22
True      0              65.08
          1              34.92
Name: is_canceled, dtype: float64

показатель оттока выше всего у группы без детей и составляет 37%